editor/src/components/navigator/layout-element-icons.ts

```diff
@@ -0,0 +1,248 @@
+import { MetadataUtils } from '../../core/model/element-metadata-utils'
+import {
+  isAnimatedElementAgainstImports,
+  isImg,
+  isImportedComponent,
+} from '../../core/model/project-file-utils'
+import {
+  isIntrinsicHTMLElement,
+  isJSXElement,
+  JSXMetadata,
+  UtopiaJSXComponent,
+} from '../../core/shared/element-template'
+import * as TP from '../../core/shared/template-path'
+import { Imports, TemplatePath } from '../../core/shared/project-file-types'
+import {
+  getOpenImportsFromState,
+  getOpenUtopiaJSXComponentsFromState,
+} from '../editor/store/editor-state'
+import { useEditorState } from '../editor/store/store-hook'
+import { isRight } from '../../core/shared/either'
+import { IcnProps } from '../../uuiui'
+import { shallowEqual } from '../../core/shared/equality-utils'
+
+interface LayoutIconResult {
+  iconProps: IcnProps
+  hasWidthOrHeight: boolean
+}
+
+export function useLayoutOrElementIcon(path: TemplatePath): LayoutIconResult {
+  return useEditorState(
+    (store) => {
+      const metadata = store.editor.jsxMetadataKILLME
+      const components = getOpenUtopiaJSXComponentsFromState(store.editor)
+      return createLayoutOrElementIconResult(path, components, metadata)
+    },
+    'useLayoutOrElementIcon',
+    (oldResult: LayoutIconResult, newResult: LayoutIconResult) => {
+      return (
+        oldResult.hasWidthOrHeight === newResult.hasWidthOrHeight ||
+        shallowEqual(oldResult.iconProps, newResult.iconProps)
+      )
+    },
+  )
+}
+
+export function useComponentIcon(path: TemplatePath): IcnProps | null {
+  return useEditorState((store) => {
+    const metadata = store.editor.jsxMetadataKILLME
+    const components = getOpenUtopiaJSXComponentsFromState(store.editor)
+    const imports = getOpenImportsFromState(store.editor)
+    return createComponentIconProps(path, components, metadata, imports)
```

Comment:
this looks like it's going to be busting the shallow equality.

Reply:
why? it's returning an object with strings and numbers

Reply:
ah yeah you are right! sorry I thought there's something more complex in there

Reply:
I changed the type here to a simple one